whisk/client.go

```diff
@@ -316,98 +319,118 @@ func BodyTruncator(body io.ReadCloser) (string, io.ReadCloser, error) {
     return string(data), reload, nil
 }
 
+func retry(attempts int, sleep time.Duration, callback func() (*http.Response, error)) (*http.Response, error)  {
+    var err error
+    var resp *http.Response
+    for i := 0; ; i++ {
+        resp, err = callback()
```

Comment:
@houshengbo do we wish to retry on all errors?  It seems that some are "soft" errors that deserve retry (like concurrency) and others are "hard" errors that will not produce different results on a retry (for example, authorization failures).

Reply:
retry shouldn't be applied blindly - a conflict (409) is an indication that something is wrong and automatically retrying may not have the desired affect in some cases.

Reply:
the logic is actually a bit subtle. i have this in a branch of the openwhisk npm -- where i've needed to use retries for a while, now.

the following status codes should never be retried:
```
statusCode === 400 || statusCode === 401 || statusCode === 404 || statusCode === 409 || statusCode === 410 || statusCode === 500
```

if the statusCode is 502... then we probably also shouldn't retry.

Reply:
we could retry on non-HTTP error code of 153, where we know we should retry...
```
Error code: 153: Concurrent modification to resource detected.
```

Reply:
153 is 409 % 256 which is the wrapping the client does to report the status code on the terminal. 

a blind retry on conflict just doesn't make sense to me: what does it mean for example to retry rule enable/disable if the concurrent operation deleted the rule, for example.

Reply:
What about checking both of the "code 409" and error message "Concurrent modification to resource detected"? Can this be limited to the case of Concurrent modification only?